src/test/regress/expected/distributed_types.out

```diff
@@ -258,7 +258,10 @@ SELECT run_command_on_workers($$SELECT row(typname, usename) FROM pg_type, pg_us
 
 -- deleting the enum cascade will remove the type from the table and the workers
 DROP TYPE te3 CASCADE;
-NOTICE:  drop cascades to column c of table t5
+ERROR:  Can't drop type "type_tests.te3" because it is used in column "c" of Citus table "t5"
```

Comment:
I do not really follow this test output.

Table t5 has an int distribution column.

```sql
CREATE TABLE t5 (a int PRIMARY KEY, b tc5[], c te3);
SELECT create_distributed_table('t5','a');
```

This looks like we disallow dropping a type that is used in any column of a distributed table. I'd expect to allow this drop operation and drop all non-distribution columns that depended on this type.

Vanilla PG allows this, and IMO so should we. Disallowing dropping any type that would cause Citus to lose information on how the data is distributed in a cluster is totally ok. Disallowing dropping a type that is not used directly or _indirectly_ in a distribution column is questionable, or debatable.

When I say _indirectly_ I mean to say that it is possible to create type_a, and a composite type_b that depends on type_a. If the user uses type_b in a distribution column, we can also disallow dropping type_a as well even though the distribution column is not type_a, but type_b.

Reply:
The problem is that when Postgres drops table columns as a result of cascading the drop from a type drop, it does this internally, hence it doesn't propagate anything to the shards. We want to maintain the same columns between the table and its shards.

>When I say indirectly I mean to say that it is possible to create type_a, and a composite type_b that depends on type_a. If the user uses type_b in a distribution column, we can also disallow dropping type_a as well even though the distribution column is not type_a, but type_b.

Good point, I hadn't thought about this. I would vote on disallowing this as well. Although we might skip this one considering it is a rare use case.

Reply:
> I prefer disallowing dropping non-distribution columns as well to maintain what I mentioned above.

I wonder what @onderkalaci thinks about this approach? Frankly, I am against this idea unless it is too much work to disallow drop type cascade only when it is used in a partition column.

Reply:
> unless it is too much work to disallow drop type cascade only when it is used in a partition column.

I don't think it is too much work.

Those columns would be left in the shards with no meaning: there is no way to reach to those columns. They might have data which would be untracked. I believe that if we are going to merge such a change, we might do it for all columns as well.

Reply:
> The problem is that when Postgres drops table columns as a result of cascading the drop from a type drop, it does this internally, hence it doesn't propagate anything to the shards.

This is unclear to me, Citus also sends DROP .. CASCADE to the workers, which should do it for the shards? 

```SQL

DROP TYPE tc1 CASCADE;
...
DETAIL:  on server onderkalaci@localhost:9705 connectionId: 6
NOTICE:  issuing DROP TYPE public.tc1 CASCADE;
...
```

> I prefer disallowing dropping non-distribution columns as well to maintain what I mentioned above.

This sounds too restrictive to me, mostly because I could not follow why is that necessary. Can you share a small repro steps that this cause problems?

Reply:
Wow, you are right. We actually have NO problem with dropping non-distribution columns.

Reply:
> Citus also sends DROP .. CASCADE to the workers

somehow I missed this. Thanks a lot, will update the PR accordingly.